x-pack/plugins/security/public/session/session_timeout.ts

```diff
@@ -0,0 +1,309 @@
+import { createSessionExpirationToast } from './session_expiration_toast';
+import type { ISessionExpired } from './session_expired';
+
+export interface SessionState extends Pick<SessionInfo, 'expiresInMs' | 'canBeExtended'> {
+  lastExtensionTime: number;
+}
+
+export class SessionTimeout {
+  private channel?: BroadcastChannelType<SessionState>;
+
+  private isVisible = document.visibilityState !== 'hidden';
+  private isFetchingSessionInfo = false;
+
+  private sessionState$ = new BehaviorSubject<SessionState>({
+    lastExtensionTime: 0,
+    expiresInMs: null,
+    canBeExtended: false,
+  });
+  private subscription?: Subscription;
+
+  private warningToast?: Toast;
+
+  private stopActivityMonitor?: Function;
+  private stopVisibilityMonitor?: Function;
+  private removeHttpInterceptor?: Function;
+
+  private stopRefreshTimer?: Function;
+  private stopWarningTimer?: Function;
+  private stopLogoutTimer?: Function;
+
+  constructor(
+    private notifications: NotificationsSetup,
+    private sessionExpired: ISessionExpired,
+    private http: HttpSetup,
+    private tenant: string
+  ) {}
+
+  public async start() {
+    if (this.http.anonymousPaths.isAnonymous(window.location.pathname)) {
+      return;
+    }
+
+    this.subscription = this.sessionState$
+      .pipe(skip(1), throttleTime(1000), tap(this.toggleEventHandlers))
+      .subscribe(this.resetTimers);
+
+    // Subscribe to a broadcast channel for session timeout messages.
+    // This allows us to synchronize the UX across tabs and avoid repetitive API calls.
+    try {
+      const { BroadcastChannel } = await import('broadcast-channel');
+      this.channel = new BroadcastChannel(`${this.tenant}/session_timeout`, {
+        webWorkerSupport: false,
+      });
+      this.channel.onmessage = this.handleChannelMessage;
+    } catch (error) {
+      // eslint-disable-next-line no-console
+      console.warn(
+        `Failed to load broadcast channel. Session management will not be kept in sync when multiple tabs are loaded.`,
+        error
+      );
+    }
+
+    return this.fetchSessionInfo();
+  }
+
+  public stop() {
+    const nextState = {
+      lastExtensionTime: 0,
+      expiresInMs: null,
+      canBeExtended: false,
+    };
+    this.toggleEventHandlers(nextState);
+    this.resetTimers(nextState);
+    this.subscription?.unsubscribe();
+    this.channel?.close();
+  }
+
+  /**
+   * Event handler that receives session information from other browser tabs.
+   */
+  private handleChannelMessage = (message: SessionState) => {
+    this.sessionState$.next(message);
+  };
+
+  /**
+   * HTTP request interceptor which ensures that API calls extend the session only if tab is
+   * visible.
+   */
+  private handleHttpRequest = (fetchOptions: HttpFetchOptionsWithPath) => {
+    // Ignore requests to external URLs
+    if (fetchOptions.path.indexOf('://') !== -1) {
+      return;
+    }
+
+    if (fetchOptions.asSystemRequest === undefined) {
```

Comment:
Yeah, I tend to agree

Reply:
I think if a developer flags a request as either a user or a system request we should to respect that. 
Otherwise we render the flag useless and should remove it completely.

Reply:
> Otherwise we render the flag useless and should remove it completely.

I think what we need is to forbid `false` value for this flag. This way API consumers won't be able to unknowingly reduce effectiveness of our security controls and can either force it to be `true` or rely on the default behavior. But, yeah, it'd probably make sense to phase it out completely eventually.